### PR TITLE
Add prompt before replacing latest release title on  prepare release run

### PR DESCRIPTION
**Pull request by chidozieononiwu**
Should resolve https://github.com/Azure/azure-sdk-tools/issues/1542

### CODE REVIEW

File: eng/common/scripts/Prepare-Release.ps1 [hunk 1]
```diff
@@ -177,8 +177,29 @@ if ($releaseTrackingOnly)
 
 if (Test-Path "Function:SetPackageVersion")
 {
-  SetPackageVersion -PackageName $packageProperties.Name -Version $newVersion -ServiceDirectory $packageProperties.ServiceDirectory -ReleaseDate $releaseDateString `
-    -PackageProperties $packageProperties
+  $changeLogEntries = Get-ChangeLogEntries -ChangeLogLocation $packageProperties.ChangeLogPath
+  $versionOfLatestEntry = $changeLogEntries[0].ReleaseStatus.Trim("()")
```
Comment (weshaggard) [inline eng/common/scripts/Prepare-Release.ps1:181]: I don't like that we are trimming this outside of the changelog script. Can you please add a helper inside of the changelog operations that can encapsulate this logic there.

File: eng/common/scripts/Prepare-Release.ps1 [hunk 2]
```diff
@@ -177,8 +177,29 @@ if ($releaseTrackingOnly)
...
+  if ($versionOfLatestEntry -as [DateTime])
+  {
+    $promptMessage = "The latest entry in the CHANGELOG.md already has a release date. Do you want to replace the latest entry title? Please enter (y or n)."
+    $promptResponse = Read-Host -Prompt $promptMessage
+    while ($promptResponse -notmatch '^[yn]$')
+    {
+      $promptResponse = Read-Host -Prompt $promptMessage
+    }
```
Comment (weshaggard) [inline eng/common/scripts/Prepare-Release.ps1:189]:
```suggestion
    while (($readInput = Read-Host -Prompt $promptMessage)  -notmatch '^[yn]$') { }
```

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 3]
```diff
@@ -351,4 +351,13 @@ function Remove-EmptySections {
     }
     $ChangeLogEntry.ReleaseContent = $sanitizedReleaseContent.ToArray()
   }
+}
+
+function Get-TrimmedVersion
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:356]: While this helper is fine I was thinking about a methot that gets the latest release date or null if unreleased.

File: eng/common/scripts/Prepare-Release.ps1 [hunk 4]
```diff
@@ -177,8 +177,24 @@ if ($releaseTrackingOnly)
...
+    SetPackageVersion -PackageName $packageProperties.Name -Version $newVersion `
+      -ServiceDirectory $packageProperties.ServiceDirectory -ReleaseDate $releaseDateString `
+      -PackageProperties $packageProperties -ReplaceLatestEntryTitle $replaceLatestEntryTitle
+  }
+  else
+  {
+    SetPackageVersion -PackageName $packageProperties.Name -Version $newVersion `
+      -ServiceDirectory $packageProperties.ServiceDirectory -ReleaseDate $releaseDateString `
+      -PackageProperties $packageProperties
+  }
```
Comment (weshaggard) [inline eng/common/scripts/Prepare-Release.ps1:197]:
```suggestion
  $releaseDate = Get-LatestReleaseDateFromChangeLog -ChangeLogLocation $packageProperties.ChangeLogPath
  $replaceLatestEntryTitle = $true
  if ($releaseDate) {
      $promptMessage = "The latest entry in the CHANGELOG.md already has a release date. Do you want to replace the latest entry or add a new entry? Please enter (y or n)."
    while (($readInput = Read-Host -Prompt $promptMessage) -notmatch '^[yn]$'){ }
    $replaceLatestEntryTitle = ($readInput -eq "y")
  }
  SetPackageVersion -PackageName $packageProperties.Name -Version $newVersion `
    -ServiceDirectory $packageProperties.ServiceDirectory -ReleaseDate $releaseDateString `
    -PackageProperties $packageProperties -ReplaceLatestEntryTitle $replaceLatestEntryTitle
```

File: eng/common/scripts/Prepare-Release.ps1 [hunk 5]
```diff
@@ -177,8 +177,23 @@ if ($releaseTrackingOnly)
...
+    while (($readInput = Read-Host -Prompt $promptMessage) -notmatch '^[yn]$'){ }
+    $replaceLatestEntryTitle = ($readInput -eq "y")
+
+    SetPackageVersion -PackageName $packageProperties.Name -Version $newVersion `
+      -ServiceDirectory $packageProperties.ServiceDirectory -ReleaseDate $releaseDateString `
+      -PackageProperties $packageProperties -ReplaceLatestEntryTitle $replaceLatestEntryTitle
+  }
+  else
+  {
+    SetPackageVersion -PackageName $packageProperties.Name -Version $newVersion `
```
Comment (weshaggard) [inline eng/common/scripts/Prepare-Release.ps1:193]: Why do we need to duplicate this SetPackageVersion call? YOu should be able to just make the call once with the correct replaceLatestEntryTitle value.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 6]
```diff
@@ -351,4 +351,22 @@ function Remove-EmptySections {
...
+  $changeLogEntries = Get-ChangeLogEntries -ChangeLogLocation $ChangeLogLocation
+  $latestVersion = $changeLogEntries[0].ReleaseStatus.Trim("()")
+  if ($latestVersion -as [DateTime])
+  {
+    return $latestVersion
+  }
+  else
+  {
+    return $null
+  }
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:371]:
```suggestion
  return ($latestVersion -as [DateTime])
```
as will return null if the type doesn't match, but if the type does match you should return the DateTime object which this does as well.

File: eng/common/scripts/Prepare-Release.ps1 [hunk 7]
```diff
@@ -177,8 +177,17 @@ if ($releaseTrackingOnly)
 
 if (Test-Path "Function:SetPackageVersion")
 {
-  SetPackageVersion -PackageName $packageProperties.Name -Version $newVersion -ServiceDirectory $packageProperties.ServiceDirectory -ReleaseDate $releaseDateString `
-    -PackageProperties $packageProperties
+  $replaceLatestEntryTitle = $true
+  $latestVersion = Get-LatestReleaseDateFromChangeLog -ChangeLogLocation $packageProperties.ChangeLogPath
+  if ($latestVersion -ne $null)
```
Comment (weshaggard) [inline eng/common/scripts/Prepare-Release.ps1:182]:
```suggestion
  if ($latestVersion)
```

tip you don't need to explicitly test for null or empty the value will result in the correct true/false value based on that.

**Review by weshaggard — APPROVED** [review]
